### ISSUE TITLE
PH_S032  + Sonartype S4457 guidance request

### ISSUE THREAD

**Issue by MalcolmEllis** [Concurrency-Lab/ParallelHelper#89]
https://rules.sonarsource.com/csharp/RSPEC-4457 recommends splitting methods with parameter checks producing outer wrapping methods that trigger PH_S032. I'm getting confused here, if I then follow PH_S032 and wrap the null check ArgumentNullException in Task.FromException, even though there's no await, is that defeating the point of separating the null checks in order to throw immediately? Apologies if this is the wrong place to ask questions like this.

**Comment by camrein** [comment]
The case you're describing is a corner case that is probably influenced by personal taste. PH_S032 intends to make the behavior of non-async methods that return a task behave similarly to methods that are async. Sonartype's S4457 makes sense in the way that they encourage fail-fast behavior if your method is misused, i.e., by passing `null` where they shouldn't. From your writing, I believe you understand the different behavior of these two variants. If not, I'm happy to explain further.

tldr; In my opinion, this case can be treated as personal preference. Since I support fail-fast behavior, I like the style S4457 encourages. I've also checked how .NET itself handles these cases. It appears to me that they do the same as S4457 suggests. For example, see [MemoryStream](https://referencesource.microsoft.com/#mscorlib/system/io/memorystream.cs,421), [FileStream](https://referencesource.microsoft.com/#mscorlib/system/io/filestream.cs,2655), and [HttpClient](https://github.com/microsoft/referencesource/blob/5697c29004a34d80acdaf5742d7e699022c64ecd/System/net/System/Net/Http/HttpClient.cs#L365).

A week ago I've improved PH_S032 with #86 to allow excluding exception types from its analysis. With this change, you should have the freedom to configure PH_S032 to not interfere with S4457. I'll probably extend it to allow excluding all types inheriting from `ArgumentException` to capture custom sub-types before the next update.

**Comment by camrein** [comment]
I've just published the update to [v3.4.0](https://github.com/Concurrency-Lab/ParallelHelper/releases/tag/v3.4.0). I hope this resolves your request. If not, feel free to open another issue.
